backend/hyper/entity.go

```diff
@@ -349,6 +352,128 @@ func (bs *Storage) LoadEntity(ctx context.Context, eid EntityID) (e *Entity, err
 	return bs.loadFromHeads(conn, eid, heads.Heads)
 }
 
+func enqueue(queue []hypersql.ChangesListFromChangeSetResult, element hypersql.ChangesListFromChangeSetResult) []hypersql.ChangesListFromChangeSetResult {
+	queue = append(queue, element) // Simply append to enqueue.
+	return queue
+}
+
+func dequeue(queue []hypersql.ChangesListFromChangeSetResult) []hypersql.ChangesListFromChangeSetResult {
+	return queue[1:] // Slice off the element once it is dequeued.
+}
+
+type changeToApply struct {
+	chID cid.Cid
+	ch   Change
+}
+
+// LoadTrustedEntity will return the lastest entity version changed by a trusted peer.
+func (bs *Storage) LoadTrustedEntity(ctx context.Context, eid EntityID) (e *Entity, err error) {
+	conn, release, err := bs.db.Conn(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer release()
+
+	defer sqlitex.Save(conn)(&err)
+
+	edb, err := hypersql.EntitiesLookupID(conn, string(eid))
+	if err != nil {
+		return nil, err
+	}
+	if edb.HDEntitiesID == 0 {
+		return nil, status.Errorf(codes.NotFound, "entity %q not found", eid)
+	}
+
+	heads, err := hypersql.ChangesGetPublicHeadsJSON(conn, edb.HDEntitiesID)
+	if err != nil {
+		return nil, err
+	}
+
+	cset, err := hypersql.ChangesResolveHeads(conn, heads.Heads)
+	if err != nil {
+		return nil, err
+	}
+
+	changes, err := hypersql.ChangesListFromChangeSet(conn, cset.ResolvedJSON, string(eid))
+
+	if len(changes) == 0 {
+		return nil, nil
+	}
+
+	lastTrustedchangeIdx := -1
+
+	buf := make([]byte, 0, 1024*1024) // preallocating 1MB for decompression.
+	/*
+		heads = queue(D)
+		for head = heads.dequeue() and head is not null:
+		  if head is not trusted:
+		    heads.enqueue(head.deps)
+		  else:
+		    break
+		for change := dequeue(changes); change != nil; dequeue(changes) {
+
+		}
+	*/
+	chToApply := []changeToApply{}
+	for idx, change := range changes {
```

Comment:
Isn't all this logic the same across all the load entity methods? Maybe we can now make it more DRY?